KAFKA-7140: Remove deprecated poll usages

viktorsomogyi opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

viktorsomogyi committed:
MINOR: use poll(Duration) in consumer example

ijuma commented:
Thanks for the PR. Maybe we can update all non-test usages of the deprecated method at once?

viktorsomogyi committed:
Replace other non-test usages as well

viktorsomogyi committed:
Removing the rest of the deprecated non-test poll usages

viktorsomogyi commented:
@ijuma thanks for the suggestion, I've updated the PR.

viktorsomogyi commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -314,7 +314,7 @@ private int maybeResetInputAndSeekToEndIntermediateTopicOffsets(final Map consum
         try (final KafkaConsumer<byte[], byte[]> client = new KafkaConsumer<>(config, new ByteArrayDeserializer(), new ByteArrayDeserializer())) {
             client.subscribe(topicsToSubscribe);
-            client.poll(1);
+            client.poll(java.time.Duration.ofMillis(1));
```
@mjsax I think you authored the change where this line was introduced. Do you know what's the reasoning for poll(1)? Did it just used the fact that it updated the metadata or were there any other reasons? I'm not sure that my change is correct here and I don't have the context.

mjsax replied:
Yes, we do this call to update the metadata. The change seems reasonable to me.

hachikuji replied:
Are you sure about  this? `poll(1)` would previously have guaranteed that the consumer joins the group. That is no longer the case with the new poll, so the call to `assignment` below is likely to return an empty set.

mjsax replied:
Because `StreamsResetter` must be the only member in the group (we check on startup that the group is empty), we can check if the assignment contains all offsets? Or we could switch to manual assignment?

hachikuji replied:
Yeah, I think manual assignment is a better option here. For one thing, it guarantees that there won't be another member in the group when you receive the assignment. It also gets around the awkward use of poll() since you do not actually want to fetch data. The basic approach would be to use the `partitionsFor` API to find the partitions for the subscribed topics and then assign them directly with `assign`.

viktorsomogyi replied:
Ok, I'll try to refactor my code this way.

viktorsomogyi commented:
Also created a jira for this as it I think it now exceeds the "minor" category :)

hachikuji commented:
@viktorsomogyi Thanks for the patch. Note that there is a semantic difference between the old and new `poll` APIs. The old API blocked if a rebalance was in progress, but the new one does not. You may need to review each usage individually to be sure that the new semantic is still appropriate.

viktorsomogyi commented:
retest this please

viktorsomogyi commented:
@hachikuji thanks for the note. Yes, I've been trying to keep in mind that poll(0) differs in this case. In some cases we could just use to the position call or poll(0) just because we make sure to have an initial value to an iterator, etc.
The only case I found so far where I'm not sure is in StreamsResetter.

hachikuji commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -388,7 +389,7 @@ object ConsoleConsumer extends Logging {
                                        consumer: Consumer[Array[Byte], Array[Byte]], val timeoutMs: Long = Long.MaxValue) {
     consumerInit()
-    var recordIter = consumer.poll(0).iterator
+    var recordIter = consumer.poll(Duration.ZERO).iterator
```
Maybe we can just initialize this to an empty iterator? The call seems intended to join the group, but that will no longer happen when the call returns.

viktorsomogyi replied:
Were you thinking about something like this?
```var recordIter = Collections.emptyList[ConsumerRecord[Array[Byte], Array[Byte]]]().iterator()```

hachikuji commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -89,9 +91,9 @@ object EndToEndLatency {
+    //a position() request is issued. Hence we need to position after we seek to ensure we see our first write.
     consumer.seekToEnd(Collections.emptyList())
-    consumer.poll(0)
+    consumer.assignment().asScala.foreach(consumer.position)
```
As far as I can tell, the assignment will be empty here. I don't think the old code worked either as intended because it did the `seekToEnd` prior to the call to `poll()`. For an end-to-end latency test, I actually wonder if we should just use manual assignment and take the group management out of the picture?

viktorsomogyi replied:
Will change to manual assignment, something like this (which seems to work).
```
val topicPartitions = consumer.listTopics().get(topic).asScala
      .map(pi => new TopicPartition(pi.topic(), pi.partition()))
      .to[List].asJava
    consumer.assign(topicPartitions)
    consumer.seekToEnd(topicPartitions)
    consumer.assignment().asScala.foreach(consumer.position)
```

hachikuji reviewed (commented):
Apologies for the delay. Left a couple comments.

viktorsomogyi committed:
Use empty iterator in ConsoleConsumer and manual assignment in EndToEndLatency

viktorsomogyi commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -88,10 +89,18 @@ object EndToEndLatency {
-    //a poll() or position() request is issued. Hence we need to poll after we seek to ensure we see our first write.
-    consumer.seekToEnd(Collections.emptyList())
-    consumer.poll(0)
+    val topics = consumer.listTopics()
```
So here it seems the old code assumed that there is a topic. Now I added a check that there must be a topic beforehand. the producer would anyway create it according to the default settings but I think that most of the latency tests executed by users won't depend on the default settings and they rather want to create topics manually according to their likings.

hachikuji replied:
I guess another approach might be to produce a dummy record to the topic first before assigning the partitions to the consumer and starting the test. That would let us retain the existing behavior.

viktorsomogyi committed:
Manual partition assignment in StreamsResetter

viktorsomogyi committed:
Replace poll calls in WorkerSinkTaskTest

viktorsomogyi committed:
Replace polls in ErrorHandlingTaskTest

viktorsomogyi commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -313,10 +316,14 @@ private int maybeResetInputAndSeekToEndIntermediateTopicOffsets(final Map consum
         try (final KafkaConsumer<byte[], byte[]> client = new KafkaConsumer<>(config, new ByteArrayDeserializer(), new ByteArrayDeserializer())) {
-            client.subscribe(topicsToSubscribe);
-            client.poll(1);
+            Map<String, List<PartitionInfo>> pi = client.listTopics();
```
Not final yet, will refactor it a bit as suggested in earlier comments.

viktorsomogyi committed:
Use partitionsFor in StreamsResetter

viktorsomogyi commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -313,10 +316,12 @@ private int maybeResetInputAndSeekToEndIntermediateTopicOffsets(final Map consum
         try (final KafkaConsumer<byte[], byte[]> client = new KafkaConsumer<>(config, new ByteArrayDeserializer(), new ByteArrayDeserializer())) {
-            client.subscribe(topicsToSubscribe);
-            client.poll(1);
+            Collection<TopicPartition> partitions = topicsToSubscribe.stream().map(client::partitionsFor)
```
@hachikuji were you thinking of something like this?

mjsax replied:
LGTM.

hachikuji replied:
Java 8 makes this much nicer than I was expecting. 😄

viktorsomogyi commented:
General update:
- some fixes around suggestions from review
- fixed test failures (some polls there needed to be replaced)
- rebased on latest trunk

viktorsomogyi commented:
retest this please

mjsax commented:
@viktorsomogyi There is a checkstyle error
```
/home/jenkins/jenkins-slave/workspace/kafka-pr-jdk10-scala2.12/core/src/main/scala/kafka/tools/StreamsResetter.java:36:8: Unused import - org.apache.kafka.common.PartitionInfo. [UnusedImports
```

hachikuji reviewed (commented):
Thanks for the updates. Just one comment for discussion.

viktorsomogyi committed:
Remove unused imports

viktorsomogyi committed:
Auto create topic in EndToEndLatency

hachikuji commented:
@viktorsomogyi Thanks for the update. LGTM overall. I think to fix the failing tests, we just need to use the right `poll()` in `WorkerSinkTaskThreadedTest`.

viktorsomogyi committed:
Fix failing tests in WorkerSinkTaskThreadedTest

viktorsomogyi commented:
Ah, indeed. If you haven't started fixing it yet, I pushed a commit with that. Let's see if we'll have any after this. :)

viktorsomogyi commented:
It seems there was a flaky afterall. I'll rerun them.

viktorsomogyi commented:
retest this please

tedyu commented:
Can this be dropped from Consumer (in trunk) ?
```
    ConsumerRecords<K, V> poll(long timeout);
```

mjsax commented:
@tedyu We cannot just remove a method from public API, as this would be a breaking change. Note, that the method is already `deprecated` though.
